Fixed BacktraceCleaner to never return an empty backtrace.

zenspider opened the pull request:
### Summary

Empty backtraces means you didn't run any code, which isn't the case,
and goes against the contract that Minitest.backtrace_cleaner expects.
This fixes a bug I've seen in a number of reports.

It would be nice if this got backported to whatever versions are
active, as this keeps coming back on minitest issues.

zenspider committed:
Fixed BacktraceCleaner to never return an empty backtrace.

Empty backtraces means you didn't run any code, which isn't the case,
and goes against the contract that Minitest.backtrace_cleaner expects.
This fixes a bug I've seen in a number of reports.

It would be nice if this got backported to whatever versions are
active, as this keeps coming back on minitest issues.

zenspider commented:
See: 

* https://github.com/seattlerb/minitest/issues/837
* https://github.com/seattlerb/minitest/issues/831

Probably others...

zenspider commented:
Also, I only ran the one test file in question. I'll monitor the CI to make sure there aren't indirect failures.

zenspider committed:
Fixes against the previous commit.

Happy to squash this before merge. Just separate so I can whittle
these down.

Most of the problem seems to be centered around the backtrace filter
being both for general use and for minitest... and I'm only aware of /
addressing minitest issues.

If they really need to be separate, then we can make a separate one
for minitest.

jonathanhefner commented on activesupport/lib/active_support/backtrace_cleaner.rb:
```diff
@@ -104,27 +104,33 @@ def add_stdlib_silencer
-        end
-
-        backtrace
+        process_collection backtrace, @silencers, :reject
```
If this is the only spot where `empty_guard` is necessary, perhaps it could be rolled into this method:

```ruby
keep = @silencers.reduce(backtrace) do |bt, s|
  bt.reject { |line| s.call(line) }
end

keep.first ? keep : backtrace
```

The above code also sidesteps the issue of Enumerator not responding to `empty?`.

zenspider replied:
I like the `.first` change... 

I gotta ask, what's the point of making a known collection lazy? You already have the collection and things like `find` and `any?` are already shortcutting. That's why I took the `lazy` out. Totally fine with putting it back.

jonathanhefner replied:
In this case, using `first`, it avoids processing the entire collection.  It looks like the `lazy` was added in #35985 as a performance optimization.  Since it was for performance reasons, it would be nice to avoid calling `first` twice, but I don't see a way around that.

jonathanhefner reviewed (commented):
I'm not sure of the bigger picture, but I left some implementation comments.

zenspider committed:
Fix failure in reporter_test.rb

I'm not sure about this one. The gist is, minitest's
UnexpectedError#message includes a backtrace but the reporter tests
here expect there not to be one in inline output.

Personally, I think that's wrong as it removes information you'd want
to debug, but this might be a philosophical difference with the rails
test ethos... or it might just be aesthetic.

I am happy to back out this change and fix the tests if you'd rather.

zenspider commented on activesupport/lib/active_support/backtrace_cleaner.rb:
```diff
@@ -104,27 +104,33 @@ def add_stdlib_silencer
-          @silencers.any? do |s|
-            s.call(line)
-          end
+        empty_guard backtrace do
+          backtrace.select { |line| @silencers.any? { |s| s.call(line) } }
```
The strictness is the question... It IS strict on minitest, but this looks to be used in multiple locations so it doesn't necessarily apply here. It might be better to have a subclass JUST for `Minitest.backtrace_filter`. I can go either way.

zenspider replied:
P.S. I lived the first 5ish years of my life in/around Lafayette!

jonathanhefner replied:
@zenspider Haha, small world!  :earth_americas:  :smile:

zenspider commented on activesupport/lib/active_support/backtrace_cleaner.rb:
```diff
@@ -104,27 +104,33 @@ def add_stdlib_silencer
-        backtrace
+      # use @filters to transform the backtrace via map
+      def filter_backtrace(backtrace)
+        process_collection backtrace, @filters, :map
```
No, not necessary... but doesn't hurt. If we pull the guard_empty down to the one below then it's a moot point and I'll unfactor the rest of these (tho, Iike the way that `process_collection` reads... thoughts?)

zenspider commented on railties/lib/rails/test_unit/reporter.rb:
```diff
@@ -17,6 +17,8 @@ def record(result)
 
       if output_inline? && result.failure && (!result.skipped? || options[:verbose])
+        result.failures.each { |f| f.backtrace.clear } if result.error?
+
```
I _really_ don't like this change. Please read the comment on the commit for my thoughts.

zenspider replied:
Given how `SuppressedSummaryReporter` works, that means when you run tests with `inline` on and you have an error, you NEVER get a backtrace? That seems bad.

tenderlove replied:
This change is causing problems in master.  I've got some failing tests in master and with this change the backtrace is gone.  Without the change I can see the backtrace.  Why do we need this change?

tenderlove commented:
Somehow this is breaking backtraces when working on Rails itself.  Before this PR, a failing test would look like this:

```
$ be ruby -I lib:test test/controller/request_forgery_protection_test.rb -n test_ignores_trailing_slash_during_generation
Run options: -n test_ignores_trailing_slash_during_generation --seed 22205

# Running:

E

Error:
PerFormTokensControllerTest#test_ignores_trailing_slash_during_generation:
NoMethodError: undefined method `have_cookie_jar?' for #<Object:0x00007f9ce19b9620>
    /Users/aaron/git/rails/actionpack/lib/action_controller/test_case.rb:522:in `ensure in process'
    /Users/aaron/git/rails/actionpack/lib/action_controller/test_case.rb:542:in `process'
    /Users/aaron/git/rails/actionpack/lib/action_controller/test_case.rb:395:in `get'
    test/controller/request_forgery_protection_test.rb:1005:in `test_ignores_trailing_slash_during_generation'


rails test test/controller/request_forgery_protection_test.rb:1004



Finished in 0.213230s, 4.6898 runs/s, 0.0000 assertions/s.
1 runs, 0 assertions, 0 failures, 1 errors, 0 skips
```

After this PR it looks like this:

```
$ be ruby -I lib:test test/controller/request_forgery_protection_test.rb -n test_ignores_trailing_slash_during_generation
Run options: -n test_ignores_trailing_slash_during_generation --seed 62892

# Running:

E

Error:
PerFormTokensControllerTest#test_ignores_trailing_slash_during_generation:
NoMethodError: undefined method `have_cookie_jar?' for #<Object:0x00007fa6d60193c8>
    


rails test test/controller/request_forgery_protection_test.rb:1004



Finished in 0.211953s, 4.7180 runs/s, 0.0000 assertions/s.
1 runs, 0 assertions, 0 failures, 1 errors, 0 skips
```

I'm investigating the issue
